### PR TITLE
Ensure ref prefix set for external_branch

### PR DESCRIPTION
**Pull request by eseliger**
This makes it easier to find a matching changeset for a given changeset spec, because we don't need to conditionally prefix external_branch. This is a step towards DB based rewirer plans.

### CODE REVIEW

File: internal/campaigns/changeset.go [hunk 1]
```diff
@@ -268,19 +269,19 @@ func (c *Changeset) SetMetadata(meta interface{}) error {
 		c.Metadata = pr
 		c.ExternalID = strconv.FormatInt(pr.Number, 10)
 		c.ExternalServiceType = extsvc.TypeGitHub
-		c.ExternalBranch = pr.HeadRefName
+		c.ExternalBranch = git.EnsureRefPrefix(pr.HeadRefName)
```
Comment (mrnugget) [inline internal/campaigns/changeset.go:272]: Can we add an assertion to the `CreateChangeset` method in the store that ensures that if `c.ExternalBranch != "" that it includes a `RefPrefix`?

File: schema/changeset_spec.schema.json [hunk 2]
```diff
@@ -30,6 +30,7 @@
         "baseRef": {
           "type": "string",
           "description": "The full name of the Git ref in the base repository that this changeset is based on (and is proposing to be merged into). This ref must exist on the base repository.",
+          "pattern": "^refs\\/heads\\/\\S+$",
```
Comment (mrnugget) [inline schema/changeset_spec.schema.json:33]: Do we send that up in `src-cli`? Or would this break clients?

File: enterprise/internal/campaigns/reconciler_test.go [hunk 3]
```diff
@@ -103,7 +103,7 @@ func TestReconcilerProcess(t *testing.T) {
 			wantChangeset: changesetAssertions{
 				publicationState: campaigns.ChangesetPublicationStatePublished,
 				externalID:       githubPR.ID,
-				externalBranch:   githubPR.HeadRefName,
+				externalBranch:   git.EnsureRefPrefix(githubPR.HeadRefName),
```
Comment (mrnugget) [inline enterprise/internal/campaigns/reconciler_test.go:106]: Can't you move that into `assertChangeset`?

File: enterprise/internal/campaigns/resolvers/changeset_connection_test.go [hunk 4]
```diff
@@ -73,7 +74,7 @@ func TestChangesetConnectionResolver(t *testing.T) {
 		repo:                repo.ID,
 		externalServiceType: "github",
 		externalID:          "12345",
-		externalBranch:      "open-pr",
+		externalBranch:      git.EnsureRefPrefix("open-pr"),
```
Comment (mrnugget) [inline enterprise/internal/campaigns/resolvers/changeset_connection_test.go:77]: I'd also move that into `createChangeset`.

**Review by mrnugget — APPROVED** [review]

File: schema/changeset_spec.schema.json (hunk 2 above)
Comment (eseliger) [reply schema/changeset_spec.schema.json:33]: Yeah we've always done that. I just wanted to make that extra clear again here.

File: enterprise/internal/campaigns/reconciler_test.go (hunk 3 above)
Comment (eseliger) [reply enterprise/internal/campaigns/reconciler_test.go:106]: I can, but I felt like making this be extra clear in the definition itself.. WDYT?

File: internal/campaigns/changeset.go (hunk 1 above)
Comment (eseliger) [reply internal/campaigns/changeset.go:272]: I added a CHECK constraint in 15c3d57e436b071548bd38f0f3d787d72762d5b2.

File: enterprise/internal/campaigns/reconciler_test.go (hunk 3 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/reconciler_test.go:106]: Then I'd rather add the `refs/heads` to the string literals.

Comment (eseliger) [reply enterprise/internal/campaigns/reconciler_test.go:106]: The thing is that github DOES return it without, and I wanted to make clear that it is what github returns, plus that we ensured the prefix is there. I can, however change it to be 

```go
"refs/heads/" + githubPR.HeadRefName,
```

if you think that's clearer

**Review by chrispine — APPROVED** [review]
LGTM

File: enterprise/internal/campaigns/reconciler_test.go (hunk 3 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/reconciler_test.go]: Yeah, I know. I just don't like the "noise" everywhere. I think in places where we use it like here, it's okay to use `git.EnsureRefPrefix`. But in places where we put a string literal in we should prefix it with `refs/heads/`

**Pull request merged**